Simple implementation of IScrollAnchorProvider.

grokys opened the pull request:
**Note: this PR targets #4040, retarget to master before merging**

## What does the pull request do?

A minimal implementation of [scroll anchoring](https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.controls.iscrollanchorprovider) which is intended to improve scrolling in the `ItemsPresenter` with items of differing heights/widths.

This functionality is defined in UWP but this PR doesn't attempt to implement all of UWP's features here, in particular:

- No equivalent of [`UIElement.CanBeScrollAnchor`](https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.uielement.canbescrollanchor#Windows_UI_Xaml_UIElement_CanBeScrollAnchor): my thoughts here are that this functionality is too niche to deserve a property being added to every control
- No equivalent of [`ScrollViewer.AnchorRequested`](https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.controls.scrollviewer.anchorrequested): this can be added later when needed

Depends on #4040

grokys committed:
Initial implementation of IScrollAnchorProvider.

On `ScrollContentPresenter`.

MarchingCube commented on src/Avalonia.Controls/Presenters/ScrollContentPresenter.cs:
```diff
@@ -133,6 +135,9 @@ public Size Viewport
         }
 
+        /// <inheritdoc/>
+        IControl IScrollAnchorProvider.CurrentAnchor => _anchor.Item1;
```
I think you could just write `_anchor.control` for clarity. Since in some places you are using `Item1` and in some `control`.

MarchingCube commented:
Does this already fix scrollbar jumping around in `ItemsRepeater` when using non-uniform items heights in `StackLayout` ?

grokys commented:
>Does this already fix scrollbar jumping around in ItemsRepeater when using non-uniform items heights in StackLayout ?

It doesn't fix the scrollbar jumping around, but it does fix scrolling with e.g. the mousewheel causing the contents of the `ItemsRepeater` jumping around. I don't think there's any way to prevent the scrollbar jumping around without having extra information about the size of non-materialized items.

MarchingCube commented:
Just so we don't forget - after this change `ItemsRepeater` sometimes gets stuck when scrolling so there might be an issue somewhere.

grokys committed:
Reset count to 0 when re-queueing control.

When a layout of a control exceeds the maximum passes and we re-queue it, reset the attempt count to 0 because otherwise it can end up stuck in the layout queue forever even when the next layout can succeed with e.g. 2 passes. This can happen with `ItemsRepeater`.

grokys committed:
Don't invalidate arrange during arrange.

When `Offset` is adjusted due to scroll ancoring during the arrange pass in `ScrollContentPresenter`, don't invalidate the arrange: we've got it handled.

grokys committed:
Use tuple name iinstead of Item1.

grokys commented:
Closing this. Will reopen a new PR with this and #4040 combined as they're hard to separate.
